### PR TITLE
Speeding up thetastar and fixing robot oscillation

### LINKED ISSUE #1938: Make a SimulatedTest for the Path Planner
### Description of the task
Currently, our path planner is tested with a set of obstacles and we evaluate the path that it generates. However, since we have a simulated test framework now, we can test how the path planner works in the "real world". This is related to #1448 in that the goal with that constant is to deal with "real world" effects. 

One way to do this is to use a SimulatedTacticTest with the MoveTactic. However, developing a new Tactic might be useful for more complicated pathing

### Acceptance criteria

- [ ] Add a set of obstacles (enemy robots?) to the field and test navigation through them
- [ ] Visually check for weird oscillations or bad behaviour
- [ ] develop validation functions to check for bad behaviour

### PR DESCRIPTION
**Pull request by nimazareian**
### Description
This PR aims to fix multiple bugs with the path planner to speed up the simulator and allow for more robots to play on the field at once.

Bugs:
1. ThetaStar path planner would search the entire field if the end coordinate was equal to the current coordinate in `findPathToEnd`. This bug was created since the current coordinate (= end coordinate) would be first added to the `closed_list` and then `visitNeighbours` was called on the rest of the coordinates, until the end coordinate was reached. As a result of the end coordinate being added to the `closed_list` at the start, it was assumed not to be the destination, and `findPathToEnd` searched the entire field not finding the destination resulting in a `nullopt` being returned. 
When this bug was created in simulated tests, the corresponding tick would take 4-8x longer to finish, which made the simulator stutter often. This happened often when there were multiple robots closed to each other and they were trying to avoid each other.
Here are the recorded max tick duration before and after adding the check for the `current_coord == end_coord`:

The rest of the collected data of tick durations before the fix can be seen in the different sheets of this   

[document](https://docs.google.com/spreadsheets/d/1Zi7qNPjtXfeFDnHpM8VKmByHwHIiCJU5euTS8XLDMoE/edit?usp=sharing)

2. In some scenarios, once a robot reached its destination, it will oscillate back and forth and never reach a steady state at the destination. The current fix for this bug is to increase the `DISTANCE_THRESHOLD` (distance from destination to be considered close enough) from `0.02` to `0.05`.  This bug should not be directly related to `ThetaStar` path planner since when the path planner was switched to `straight_line_path_planner` the oscillation continued. The problem may be related to the motion controller, however this theory needs to be tested.

Some GIFs of both the bugs can also be seen [here](https://docs.google.com/document/d/1HrwoBAHb9rMVOmqNTNw5lzPWGBVy1JzYZZG9EBZgy8o/edit?usp=sharing).

`Log(INFO)` was also added to simulated test fixture to print some data about the tick durations for each test, which should help identify any other problems that could be slowing down the simulation.

### Testing Done
Added `simulated_theta_star_test.cpp`
   - Side feature: added a zigzag simulated test similar to the 2021 SSL dribbling hardware challenge (Only tests the path planner and does not actually dribble a ball). 

### Resolved Issues
Resolves #1938 
Related to #1448 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    motion_constraints.insert(MotionConstraint::ENEMY_ROBOTS_COLLISION);
+    setMotionConstraints(motion_constraints);
+
+    std::vector<ValidationFunction> terminating_validation_functions = {
+        [destination, tactic](std::shared_ptr<World> world_ptr,
+                              ValidationCoroutine::push_type& yield) {
+            while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(9.5))
+            {
+                yield("Timestamp not at 9.5s");
+            }
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Can we check that the robot stabilizes to a single location? Might as well do the same for all the other theta star path planner tests

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 2]
```diff
@@ -179,7 +189,23 @@ void SimulatedTestFixture::runTest(
             continue;
         }
         validation_functions_done = tickTest(simulation_time_step, ai_time_step, world);
+
+        // Calculate tick durations
+        double duration_ms = ::TestUtil::millisecondsSince(start_tick_time);
+        max_tick_duration =
+            max_tick_duration > duration_ms ? max_tick_duration : duration_ms;
+        min_tick_duration =
+            min_tick_duration < duration_ms ? min_tick_duration : duration_ms;
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: consider using std::max and std::min

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 3]
```diff
@@ -179,7 +189,23 @@ void SimulatedTestFixture::runTest(
...
+        tick_count++;
     }
+    // Output the tick duration results
+    double total_simulation_duration = ::TestUtil::millisecondsSince(start_time);
+    double avg_tick_duration         = total_simulation_duration / (double)tick_count;
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: Can we sum up tick durations similar to how we calculate min/max to get a more accurate number? Seems like we wouldn't need to initialize tick count to 2 and have this note: `value is not accurate if stop_ai_on_start flag was used` (Also, use static_cast<double> instead of (double))

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            robotInPolygon(1, expectedFinalPosition, world_ptr, yield);
+        }};
+
+    std::vector<ValidationFunction> non_terminating_validation_functions = {};
+
+    runTest(terminating_validation_functions, non_terminating_validation_functions,
+            Duration::fromSeconds(15));
+}
+
+TEST_F(SimulatedThetaStarTest, test_theta_star_zig_zag_test)
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:160]: very cool test

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    std::set<MotionConstraint> motion_constraints;
+    motion_constraints.insert(MotionConstraint::ENEMY_ROBOTS_COLLISION);
+    setMotionConstraints(motion_constraints);
+
+
+    std::vector<ValidationFunction> terminating_validation_functions = {
+        [destination, tactic](std::shared_ptr<World> world_ptr,
+                              ValidationCoroutine::push_type& yield) {
+            Rectangle expectedFinalPosition(Point(2.015, 1.015), Point(1.985, 0.985));
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: snake_case please and similarly for all the instances

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    std::vector<ValidationFunction> non_terminating_validation_functions = {};
+
+    runTest(terminating_validation_functions, non_terminating_validation_functions,
+            Duration::fromSeconds(15));
+}
+
+TEST_F(SimulatedThetaStarTest, test_theta_no_obstacle_straight_path)
+{
+    Point destination      = Point(-2, 1);
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:128]: It seems like this test passes immediately? 
I'm seeing max tick duration: 0ms, min tick duration: 10000ms, and avg tick duration: 0.0004205ms

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 7]
```diff
@@ -291,7 +293,9 @@ bool ThetaStarPathPlanner::findPathToEnd(const Coordinate &end_coord)
         // Add this vertex to the closed list
         closed_list.insert(current_coord);
 
-        if (visitNeighbours(current_coord, end_coord))
+        // Check if current coordinate or its surrounding neighbours are at the
+        // destination
+        if (current_coord == end_coord || visitNeighbours(current_coord, end_coord))
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: is this change still necessary?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Great to see this path planner bug getting fixed and nice tests!

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 1 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Would that be a new terminating validation function similar to `robot_in_polygon_validation` which also checks that the velocity is 0 once its in the polygon? Or would it look different?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 7 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: You're correct, it's not needed anymore. Updated

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 6 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:128]: Updated!

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 5 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Updated!

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 3 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: Good idea, updated.

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 2 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: Updated!

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Yeah that makes sense; we'd like to check this validation function over hundreds of ticks to make sure it stays stopped.

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 8]
```diff
@@ -169,8 +169,18 @@ void SimulatedTestFixture::runTest(
 
     // Tick one frame to aid with visualization
     bool validation_functions_done = tickTest(simulation_time_step, ai_time_step, world);
+
+    // Logging duration of each tick starting at second tick
+    int tick_count             = 0;
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: nit:
```suggestion
    unsigned int tick_count             = 0;
```

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp [hunk 9]
```diff
@@ -6,6 +6,7 @@
 #include "software/ai/navigator/obstacle/obstacle.h"
 #include "software/ai/navigator/obstacle/robot_navigation_obstacle_factory.h"
 #include "software/geom/point.h"
+#include "software/test_util/test_util.h"
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: nit: this looks unnecessary

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 10]
```diff
@@ -169,8 +169,18 @@ void SimulatedTestFixture::runTest(
 
     // Tick one frame to aid with visualization
     bool validation_functions_done = tickTest(simulation_time_step, ai_time_step, world);
+
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp:174]: we can use this `tickTest` to initialize `max_tick_duration` and `min_tick_duration` as well as update `total_tick_duration` and `tick_count`

**Review by jordanzliu — COMMENTED** [review]
LGTM other than some minor code quality stuff

**Review by itsarune — COMMENTED** [review]
great simulated_theta_start_test! nothing to add beyond what's mentioned so far

**Review by akhilveeraghanta — COMMENTED** [review]
not much to add beyond what has already been said, nice work!

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 10 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp:174]: Could you elaborate on your recommendation? How would the values be updated inside `tickTest` and then printed in `runTest`? Would the values be passed as an argument by reference to `tickTest` when it's called so it can be updated there? If that's the case, what would be the benefit of doing such

Comment (jonathanlew) [reply src/software/simulated_tests/simulated_test_fixture.cpp:174]: what I mean is that instead of setting min_tick_duration to 10000.0, we can set min_tick_duration to the time that the tickTest on line 171 takes to run

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 8 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: Updated!

File: src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp (hunk 9 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/theta_star_path_planner_test.cpp]: You're right, removed

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 10 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp:174]: I see, that makes sense! I've updated it

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 1 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Added a new validation function. Let me know if you think anything should be changed

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+class SimulatedThetaStarTest : public SimulatedTacticTestFixture
+{
+};
+
+TEST_F(SimulatedThetaStarTest, test_theta_star_robot_and_dest_in_obstacle)
+{
+    // Both initial_position and destination are in obstacle and in the same coordinate
+    Point destination      = Point(1, 0);
+    Point initial_position = Point(0.96905413818359376, -0.26277551269531252);
```
[the remark this answers is not in the document]
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:25]: The values where taken from a specific case that used to fail. I've updated the comments to be more descriptive

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+    std::set<MotionConstraint> motion_constraints;
+    motion_constraints.insert(MotionConstraint::ENEMY_ROBOTS_COLLISION);
+    setMotionConstraints(motion_constraints);
+
+
+    std::vector<ValidationFunction> terminating_validation_functions = {
+        [destination, tactic](std::shared_ptr<World> world_ptr,
+                              ValidationCoroutine::push_type& yield) {
+            Rectangle expected_final_position(Point(0.985, -0.26), Point(1.15, -0.3));
```
[the remark this answers is not in the document]
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:46]: Comment added

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
```
[the remark this answers is not in the document]
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:81]: Comment added

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 14]
```diff
@@ -167,10 +167,23 @@ void SimulatedTestFixture::runTest(
...
+    unsigned int tick_count    = 1;
+    double duration_ms         = ::TestUtil::millisecondsSince(start_tick_time);
+    double total_tick_duration = duration_ms;
+    double max_tick_duration   = duration_ms;
+    double min_tick_duration   = duration_ms;
+
     while (simulator->getTimestamp() < timeout_time && !validation_functions_done)
     {
+        // Record starting time
+        start_tick_time = std::chrono::system_clock::now();
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: this should be after we sleep, i.e. right before tickTest

**Review by jonathanlew — COMMENTED** [review]
looking good, please fix ci and one comment

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 14 above)
Comment (nimazareian) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: Updated

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:46]: Just gonna follow up on this. Why this rectangle? Why doesn't this rectangle contain the destination (1,0)?

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:81]: Why this rectangle? Why doesn't this rectangle contain the destination (1,0)?

**Review by jonathanlew — COMMENTED** [review]
Just following up on one of @jordanzliu 's comments

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp:81]: Since the enemy robot is at (1, 0) and contains the destination with in it, the friendly robot avoids ramming into the enemy robot by going to the nearest free point which ends up being below the enemy robot.

File: src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    motion_constraints.insert(MotionConstraint::ENEMY_ROBOTS_COLLISION);
+    setMotionConstraints(motion_constraints);
+
+
+    std::vector<ValidationFunction> terminating_validation_functions = {
+        [destination, tactic](std::shared_ptr<World> world_ptr,
+                              ValidationCoroutine::push_type& yield) {
+            // Small rectangle around the destination point that the robot should be
+            // stationary within for 15 ticks
+            Rectangle expected_final_position(Point(0.985, -0.26), Point(1.15, -0.3));
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: this comment isn't accurate since the rectangle doesn't contain the destination

Comment (nimazareian) [reply src/software/ai/navigator/path_planner/simulated_theta_star_test.cpp]: Good point, I have updated it.

**Review by jonathanlew — APPROVED** [review]
LGTM! :rocket:

**Review by akhilveeraghanta — APPROVED** [review]
Nice job w/ the methodological approach and tackling this! those are some significant performance gains :100:
